### PR TITLE
image_pull: check image already pulled

### PR DESCRIPTION
**Pull request by runcom**
This is an optimization of our image pull code path. It's basically
how docker handles pulls as well. Let's be smart and check the image in
pull code path as well.

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

**Comment by runcom** [comment]
@mrunalp @sameo PTAL

**Comment by mrunalp** [comment]
:+1: will wait for the e2e tests

**Comment by sameo** [comment]
LGTM

**Comment by runcom** [comment]
retest

**Comment by runcom** [comment]
This is broken this way, we need to still return an unauthorized error if we're not authorized (even if the image is in the node already, that's docker behavior and node-e2e test failing relies on that).

**Comment by runcom** [comment]
@sameo @mrunalp PTAL this now matches docker behavior where before looking into local storage, it does some preliminary registry check and fail if any fails. Should be ok now (lucky enough, CI caught this! :tada:)

File: server/image_pull.go [hunk 1]
```diff
@@ -36,7 +37,11 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (*pb.P
 			}
 		}
 	}
-	options := &copy.Options{}
+	options := &copy.Options{
+		// TODO: we need a way to specify insecure registries like docker
+		//DockerInsecureSkipTLSVerify: true,
```
Comment (runcom) [inline server/image_pull.go]: This is here https://github.com/kubernetes-incubator/cri-o/pull/577

**Comment by runcom** [comment]
Please review and possibly merge https://github.com/kubernetes-incubator/cri-o/pull/577 before this one.

**Comment by runcom** [comment]
> 63/63 passed on RHEL - Passed.
> 120/121 k8s node-e2e passed on RHEL - Failed.
> 
> 26/27 passed on Fedora - Failed.
> 119/121 k8s node-e2e passed on Fedora - Failed.
> 
> 63/63 passed on CentOS - Passed.
> 0/0 k8s node-e2e passed on CentOS - Failed.
> 
> Log - https://aos-ci.s3.amazonaws.com/kubernetes-incubator/cri-o/crio-integration-tests-prs/454/fullresults.txt

old run, new one should be coming.

**Comment by runcom** [comment]
failure in RHEL is network related (/etc/hosts) this should be good to go

**Comment by mrunalp** [comment]
LGTM

**Pull request merged**